### PR TITLE
[JENKINS-16063] - Inject PROMOTED_USER_ID variable + minor issue fixes

### PR DESCRIPTION
**Pull request by oleg-nenashev**
It's an update of #52

https://issues.jenkins-ci.org/browse/JENKINS-16063

- [x] Inject PROMOTED_USER_ID
- [x] Properly handle <code>UserIdCause</code> and the legacy <code>UserCause</code>
- [x] Avoid NPEs if ManualCondition.Badge getName()/GetId() return null somehow
- [x] Some tests from #52

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
@reviewbybees  ready for the review

File: src/main/java/hudson/plugins/promoted_builds/Promotion.java [hunk 1]
```diff
@@ -127,24 +131,76 @@ public EnvVars getEnvironment(TaskListener listener) throws IOException, Interru
...
+        //fallback to badge lookup for compatibility 
+        for (PromotionBadge badget : getStatus().getBadges()) {
+            if (badget instanceof ManualCondition.Badge) {
+                final String nameFromBadge = ((ManualCondition.Badge) badget).getUserName();
+                if (nameFromBadge != null) {
+                    return nameFromBadge;
+                }
+            }
+        }
+        return "anonymous";
```
Comment (amuniz) [inline src/main/java/hudson/plugins/promoted_builds/Promotion.java]: ```Jenkins.ANONYMOUS.getName()```?

File: src/main/java/hudson/plugins/promoted_builds/conditions/ManualCondition.java [hunk 2]
```diff
@@ -239,6 +239,14 @@ public String getUserName() {
         }
 
         @Exported
+        public String getUserId() {
+            if (authenticationName == null)
```
Comment (amuniz) [inline src/main/java/hudson/plugins/promoted_builds/conditions/ManualCondition.java]: God kills a kitty every time a developer writes an "if" without braces.

**Comment by amuniz** [comment]
:bee: (only minor comments)

File: src/main/java/hudson/plugins/promoted_builds/conditions/ManualCondition.java (hunk 2 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/promoted_builds/conditions/ManualCondition.java]: Thanks for pointing it!
Actually this method has been loaned from the original PR. authenticationName in getUserId() seems to be an issue, I need to review it

**Comment by oleg-nenashev** [comment]
@amuniz Addressed all comments

**Comment by amuniz** [comment]
:bee:

**Comment by oleg-nenashev** [comment]
Rebuilding after the test fix

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Pull request merged**